File: CS/GridGroupSelect_sl/dxExample/GroupChildSelector.cs
using DevExpress.Xpf.Grid;
using System.Windows;
using System.Windows.Input;

namespace dxExample {
    public class GroupChildSelector : DependencyObject {
        static readonly DependencyProperty ModeProperty = DependencyProperty.RegisterAttached("Mode", typeof(ChildSelectionMode), typeof(GroupChildSelector), new PropertyMetadata(ChildSelectionMode.None, new PropertyChangedCallback(OnModeChanged)));

        public static ChildSelectionMode GetMode(DependencyObject obj) {
            return (ChildSelectionMode)obj.GetValue(ModeProperty);
        }
        public static void SetMode(DependencyObject obj, ChildSelectionMode value) {
            obj.SetValue(ModeProperty, value);
        }

        static void OnModeChanged(DependencyObject d, DependencyPropertyChangedEventArgs e) {
            if(!(d is TableView)) return;
            TableView view = (d as TableView);
            if(!view.IsLoaded) {
                view.Loaded += OnViewLoaded;
            } else {
                SubscribeViewEvents(view);
            }
        }

        static void OnViewLoaded(object sender, RoutedEventArgs e) {
            TableView view = (sender as TableView);
            SubscribeViewEvents(view);
        }
        static void SubscribeViewEvents(TableView view) {
            view.AddHandler(TableView.MouseLeftButtonUpEvent, new MouseButtonEventHandler(OnMouseLeftButtonUp), true);
            view.Grid.AddHandler(GridControl.GroupRowExpandingEvent, new RowAllowEventHandler(OnGroupRowExpanding));
        }
        static void OnMouseLeftButtonUp(object sender, MouseButtonEventArgs e) {
            TableView view = (sender as TableView);
            TableViewHitInfo hitInfo = view.CalcHitInfo(e.OriginalSource as DependencyObject);
            if(hitInfo.InRow && view.Grid.IsGroupRowHandle(hitInfo.RowHandle)) {
                view.BeginSelection();
                SelectChild(view, hitInfo.RowHandle);
                view.EndSelection();
            }
        }
        static void OnGroupRowExpanding(object sender, RowAllowEventArgs e) {
            TableView view = (e.Source as TableView);
            view.BeginSelection();
            SelectChild(view, e.RowHandle);
            view.EndSelection();
        }
        static void SelectChild(TableView view, int groupRowHandle) {
            int childRowCount = view.Grid.GetChildRowCount(groupRowHandle);
            view.BeginSelection();
            for(int i = 0; i < childRowCount; i++) {
                int childRowHandle = view.Grid.GetChildRowHandle(groupRowHandle, i);
                if(GetMode(view) == ChildSelectionMode.Hierarchical && view.Grid.IsGroupRowHandle(childRowHandle) && view.Grid.IsGroupRowExpanded(childRowHandle)) {
                    SelectChild(view, childRowHandle);
                }
                view.SelectRow(childRowHandle);
            }
            view.EndSelection();
        }
    }

    public enum ChildSelectionMode {
        None,
        Child,
        Hierarchical,
    }
}

File: VB/GridGroupSelect_sl_vb/dxExample/GroupChildSelector.vb
﻿Imports Microsoft.VisualBasic
Imports DevExpress.Xpf.Grid
Imports System.Windows
Imports System.Windows.Input

Namespace dxExample
	Public Class GroupChildSelector
		Inherits DependencyObject
		Private Shared ReadOnly ModeProperty As DependencyProperty = DependencyProperty.RegisterAttached("Mode", GetType(ChildSelectionMode), GetType(GroupChildSelector), New PropertyMetadata(ChildSelectionMode.None, New PropertyChangedCallback(AddressOf OnModeChanged)))

		Public Shared Function GetMode(ByVal obj As DependencyObject) As ChildSelectionMode
			Return CType(obj.GetValue(ModeProperty), ChildSelectionMode)
		End Function
		Public Shared Sub SetMode(ByVal obj As DependencyObject, ByVal value As ChildSelectionMode)
			obj.SetValue(ModeProperty, value)
		End Sub

		Private Shared Sub OnModeChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
			If Not(TypeOf d Is TableView) Then
				Return
			End If
			Dim view As TableView = (TryCast(d, TableView))
			If (Not view.IsLoaded) Then
				AddHandler view.Loaded, AddressOf OnViewLoaded
			Else
				SubscribeViewEvents(view)
			End If
		End Sub

		Private Shared Sub OnViewLoaded(ByVal sender As Object, ByVal e As RoutedEventArgs)
			Dim view As TableView = (TryCast(sender, TableView))
			SubscribeViewEvents(view)
		End Sub
		Private Shared Sub SubscribeViewEvents(ByVal view As TableView)
			view.AddHandler(TableView.MouseLeftButtonUpEvent, New MouseButtonEventHandler(AddressOf OnMouseLeftButtonUp), True)
			view.Grid.AddHandler(GridControl.GroupRowExpandingEvent, New RowAllowEventHandler(AddressOf OnGroupRowExpanding))
		End Sub
		Private Shared Sub OnMouseLeftButtonUp(ByVal sender As Object, ByVal e As MouseButtonEventArgs)
			Dim view As TableView = (TryCast(sender, TableView))
			Dim hitInfo As TableViewHitInfo = view.CalcHitInfo(TryCast(e.OriginalSource, DependencyObject))
			If hitInfo.InRow AndAlso view.Grid.IsGroupRowHandle(hitInfo.RowHandle) Then
				view.BeginSelection()
				SelectChild(view, hitInfo.RowHandle)
				view.EndSelection()
			End If
		End Sub
		Private Shared Sub OnGroupRowExpanding(ByVal sender As Object, ByVal e As RowAllowEventArgs)
			Dim view As TableView = (TryCast(e.Source, TableView))
			view.BeginSelection()
			SelectChild(view, e.RowHandle)
			view.EndSelection()
		End Sub
		Private Shared Sub SelectChild(ByVal view As TableView, ByVal groupRowHandle As Integer)
			Dim childRowCount As Integer = view.Grid.GetChildRowCount(groupRowHandle)
			view.BeginSelection()
			For i As Integer = 0 To childRowCount - 1
				Dim childRowHandle As Integer = view.Grid.GetChildRowHandle(groupRowHandle, i)
				If GetMode(view) = ChildSelectionMode.Hierarchical AndAlso view.Grid.IsGroupRowHandle(childRowHandle) AndAlso view.Grid.IsGroupRowExpanded(childRowHandle) Then
					SelectChild(view, childRowHandle)
				End If
				view.SelectRow(childRowHandle)
			Next i
			view.EndSelection()
		End Sub
	End Class

	Public Enum ChildSelectionMode
		None
		Child
		Hierarchical
	End Enum
End Namespace
